From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Sat, 1 Jan 2022 21:24:50 -0800
Subject: [PATCH] Fix saving configs with more long comments

This is a bug with snakeyaml
PR: https://bitbucket.org/snakeyaml/snakeyaml/pull-requests/3
Issue: https://bitbucket.org/snakeyaml/snakeyaml/issues/518/comments-could-cause-queue-full

Added the entire Emitter class from snakeyaml because
dev-imports doesn't work with non-mojang-added dependencies

Replacement for upstream: https://hub.spigotmc.org/stash/projects/SPIGOT/repos/bukkit/commits/a7505b3cd0498baca152777767f0e4ddebbe4d1a

diff --git a/src/main/java/org/yaml/snakeyaml/emitter/Emitter.java b/src/main/java/org/yaml/snakeyaml/emitter/Emitter.java
index 021cb0e706657dc86f67e5a30f9947e7a891f5df..ca32c00775f3d98abea39bbfeb0268bb9efbc12b 100644
--- a/src/main/java/org/yaml/snakeyaml/emitter/Emitter.java
+++ b/src/main/java/org/yaml/snakeyaml/emitter/Emitter.java
@@ -45,6 +45,7 @@ import org.yaml.snakeyaml.util.ArrayStack;
 
 import java.io.IOException;
 import java.io.Writer;
+import java.util.ArrayDeque;
 import java.util.HashMap;
 import java.util.HashSet;
 import java.util.Iterator;
@@ -173,7 +174,7 @@ public final class Emitter implements Emitable {
     // Scalar analysis and style.
     private ScalarAnalysis analysis;
     private DumperOptions.ScalarStyle style;
-    
+
     // Comment processing
     private final CommentEventsCollector blockCommentsCollector;
     private final CommentEventsCollector inlineCommentsCollector;
@@ -187,7 +188,8 @@ public final class Emitter implements Emitable {
         this.states = new ArrayStack<EmitterState>(100);
         this.state = new ExpectStreamStart();
         // Current event and the event queue.
-        this.events = new ArrayBlockingQueue<Event>(100);
+        this.events = new ArrayDeque<>(100); // Paper - allow more than 100 events (or comments)
+        // this.events = new ArrayBlockingQueue<>(100);
         this.event = null;
         // The current indentation level and the stack of previous indents.
         this.indents = new ArrayStack<Integer>(10);
@@ -240,12 +242,12 @@ public final class Emitter implements Emitable {
         // Scalar analysis and style.
         this.analysis = null;
         this.style = null;
-        
+
         // Comment processing
         this.blockCommentsCollector = new CommentEventsCollector(events,
-                CommentType.BLANK_LINE, CommentType.BLOCK);
+            CommentType.BLANK_LINE, CommentType.BLOCK);
         this.inlineCommentsCollector = new CommentEventsCollector(events,
-                CommentType.IN_LINE);
+            CommentType.IN_LINE);
     }
 
     public void emit(Event event) throws IOException {
@@ -288,7 +290,7 @@ public final class Emitter implements Emitable {
         }
         return false;
     }
-    
+
     private boolean needEvents(Iterator<Event> iter, int count) {
         int level = 0;
         int actualCount = 0;
@@ -384,9 +386,9 @@ public final class Emitter implements Emitable {
                     }
                 }
                 boolean implicit = first && !ev.getExplicit() && !canonical
-                        && ev.getVersion() == null
-                        && (ev.getTags() == null || ev.getTags().isEmpty())
-                        && !checkEmptyDocument();
+                    && ev.getVersion() == null
+                    && (ev.getTags() == null || ev.getTags().isEmpty())
+                    && !checkEmptyDocument();
                 if (!implicit) {
                     writeIndent();
                     writeIndicator("---", true, false, false);
@@ -456,14 +458,14 @@ public final class Emitter implements Emitable {
                 expectScalar();
             } else if (event instanceof SequenceStartEvent) {
                 if (flowLevel != 0 || canonical || ((SequenceStartEvent) event).isFlow()
-                        || checkEmptySequence()) {
+                    || checkEmptySequence()) {
                     expectFlowSequence();
                 } else {
                     expectBlockSequence();
                 }
             } else {// MappingStartEvent
                 if (flowLevel != 0 || canonical || ((MappingStartEvent) event).isFlow()
-                        || checkEmptyMapping()) {
+                    || checkEmptyMapping()) {
                     expectFlowMapping();
                 } else {
                     expectBlockMapping();
@@ -762,7 +764,7 @@ public final class Emitter implements Emitable {
         ScalarStyle style = scalarEvent.getScalarStyle();
         return style == ScalarStyle.FOLDED || style == ScalarStyle.LITERAL;
     }
-    
+
     private class ExpectBlockMappingSimpleValue implements EmitterState {
         public void expect() throws IOException {
             writeIndicator(":", false, false, false);
@@ -821,7 +823,7 @@ public final class Emitter implements Emitable {
         if (event instanceof ScalarEvent) {
             ScalarEvent e = (ScalarEvent) event;
             return e.getAnchor() == null && e.getTag() == null && e.getImplicit() != null && e
-                    .getValue().length() == 0;
+                .getValue().length() == 0;
         }
         return false;
     }
@@ -853,8 +855,8 @@ public final class Emitter implements Emitable {
             length += analysis.getScalar().length();
         }
         return length < maxSimpleKeyLength && (event instanceof AliasEvent
-                || (event instanceof ScalarEvent && !analysis.isEmpty() && !analysis.isMultiline())
-                || checkEmptySequence() || checkEmptyMapping());
+            || (event instanceof ScalarEvent && !analysis.isEmpty() && !analysis.isMultiline())
+            || checkEmptySequence() || checkEmptyMapping());
     }
 
     // Anchor, Tag, and Scalar processors.
@@ -881,8 +883,8 @@ public final class Emitter implements Emitable {
                 style = chooseScalarStyle();
             }
             if ((!canonical || tag == null) && ((style == null && ev.getImplicit()
-                    .canOmitTagInPlainScalar()) || (style != null && ev.getImplicit()
-                    .canOmitTagInNonPlainScalar()))) {
+                .canOmitTagInPlainScalar()) || (style != null && ev.getImplicit()
+                .canOmitTagInNonPlainScalar()))) {
                 preparedTag = null;
                 return;
             }
@@ -918,7 +920,7 @@ public final class Emitter implements Emitable {
         }
         if (ev.isPlain() && ev.getImplicit().canOmitTagInPlainScalar()) {
             if (!(simpleKeyContext && (analysis.isEmpty() || analysis.isMultiline()))
-                    && ((flowLevel != 0 && analysis.isAllowFlowPlain()) || (flowLevel == 0 && analysis.isAllowBlockPlain()))) {
+                && ((flowLevel != 0 && analysis.isAllowFlowPlain()) || (flowLevel == 0 && analysis.isAllowBlockPlain()))) {
                 return null;
             }
         }
@@ -1131,8 +1133,8 @@ public final class Emitter implements Emitable {
             }
             if (!(c == '\n' || (0x20 <= c && c <= 0x7E))) {
                 if (c == 0x85 || (c >= 0xA0 && c <= 0xD7FF)
-                        || (c >= 0xE000 && c <= 0xFFFD)
-                        || (c >= 0x10000 && c <= 0x10FFFF)) {
+                    || (c >= 0xE000 && c <= 0xFFFD)
+                    || (c >= 0x10000 && c <= 0x10FFFF)) {
                     // unicode is used
                     if (!this.allowUnicode) {
                         specialCharacters = true;
@@ -1220,7 +1222,7 @@ public final class Emitter implements Emitable {
         }
 
         return new ScalarAnalysis(scalar, false, lineBreaks, allowFlowPlain, allowBlockPlain,
-                allowSingleQuoted, allowBlock);
+            allowSingleQuoted, allowBlock);
     }
 
     // Writers.
@@ -1320,7 +1322,7 @@ public final class Emitter implements Emitable {
             if (spaces) {
                 if (ch == 0 || ch != ' ') {
                     if (start + 1 == end && this.column > this.bestWidth && split && start != 0
-                            && end != text.length()) {
+                        && end != text.length()) {
                         writeIndent();
                     } else {
                         int len = end - start;
@@ -1379,7 +1381,7 @@ public final class Emitter implements Emitable {
                 ch = text.charAt(end);
             }
             if (ch == null || "\"\\\u0085\u2028\u2029\uFEFF".indexOf(ch) != -1
-                    || !('\u0020' <= ch && ch <= '\u007E')) {
+                || !('\u0020' <= ch && ch <= '\u007E')) {
                 if (start < end) {
                     int len = end - start;
                     this.column += len;
@@ -1418,7 +1420,7 @@ public final class Emitter implements Emitable {
                 }
             }
             if ((0 < end && end < (text.length() - 1)) && (ch == ' ' || start >= end)
-                    && (this.column + (end - start)) > this.bestWidth && split) {
+                && (this.column + (end - start)) > this.bestWidth && split) {
                 String data;
                 if (start >= end) {
                     data = "\\";
@@ -1443,7 +1445,7 @@ public final class Emitter implements Emitable {
         }
         writeIndicator("\"", false, false, false);
     }
-    
+
     private boolean writeCommentLines(List<CommentLine> commentLines) throws IOException {
         boolean wroteComment = false;
         if(emitComments) {
@@ -1470,7 +1472,7 @@ public final class Emitter implements Emitable {
         }
         return wroteComment;
     }
-    
+
     private void writeBlockComment() throws IOException {
         if(!blockCommentsCollector.isEmpty()) {
             writeIndent();
